### PR TITLE
Sync Newest Simplex Updates

### PR DESCRIPTION
**Pull request by samliok**
## Why this should be merged

Simplex repo has had some changes that break the API in avalanchego. This PR syncs those changes

### CODE REVIEW

File: simplex/storage.go [hunk 1]
```diff
@@ -228,6 +247,23 @@ func (s *Storage) retrieveFinalization(seq uint64) (simplex.Finalization, error)
 	return canotoFinalization.toFinalization(s.deserializer)
 }
 
+func (s *Storage) retrieveBlacklist(seq uint64) (simplex.Blacklist, error) {
+	blacklistBytes, err := s.db.Get(blacklistKey(seq))
+	if err != nil {
+		if err == database.ErrNotFound {
```
Comment (DracoLi) [inline simplex/storage.go:255]: nit: using `errors.Is` for safety if error ever gets wrapped.
```suggestion
		if errors.Is(err, database.ErrNotFound) {
```

File: simplex/block_test.go [hunk 2]
```diff
@@ -75,6 +75,21 @@ func TestBlockSerialization(t *testing.T) {
 				return nil, nil
 			},
 		},
+		{
```
Comment (DracoLi) [inline simplex/block_test.go:84]: should we add a test case of the happy path? serializing block bytes with a valid non-empty blacklist?

File: simplex/block_test.go [hunk 3]
```diff
@@ -75,6 +75,42 @@ func TestBlockSerialization(t *testing.T) {
...
+				}
+				return cBlock.MarshalCanoto()
+			}(),
+			expectedError: errFailedToParseBlacklist,
+			parseFunc: func(_ context.Context, _ []byte) (snowman.Block, error) {
+				return testBlock, nil
+			},
+		},
+		{
+			name: "valid blacklist",
```
Comment (yacovm) [inline simplex/block_test.go]: why do we need a test for a valid blacklist? Can't we add a blacklist to the block in line 30:

```
	b := &Block{
		vmBlock: testBlock,
		metadata: simplex.ProtocolMetadata{
			Version: 1,
			Epoch:   1,
			Round:   1,
			Seq:     1,
			Prev:    genesisBlock.digest,
		},
		blacklist: simplex.Blacklist{
			NodeCount: 3,
			SuspectedNodes: []simplex.SuspectedNode{
				{NodeIndex: 0, SuspectingCount: 2, OrbitSuspected: 11},
			},
		},
	}
```

And now we cover this too?

File: simplex/qc.go [hunk 4]
```diff
@@ -199,6 +199,24 @@ func (a *SignatureAggregator) Aggregate(signatures []simplex.Signature) (simplex
 	}, nil
 }
 
+// IsQuorum checks if the provided nodes are a quorum of the membership set.
+// For now, this is calculated using one node = one vote, but in the future we can adjust
+// this calculation to cross reference validator weights if we want to support PoS.
+func (a *SignatureAggregator) IsQuorum(nodes []simplex.NodeID) bool {
```
Comment (yacovm) [inline simplex/qc.go:205]: This function isn't tested.

File: simplex/storage.go [hunk 5]
```diff
@@ -182,7 +195,13 @@ func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finali
 func finalizationKey(seq uint64) []byte {
 	seqBuff := make([]byte, 8)
 	binary.BigEndian.PutUint64(seqBuff, seq)
-	return seqBuff
+	return append(finalizationPrefix, seqBuff...)
```
Comment (yacovm) [inline simplex/storage.go]: can we allocate enough bytes so we don't need to do this append?

File: simplex/storage.go [hunk 6]
```diff
@@ -169,6 +177,11 @@ func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finali
 		return fmt.Errorf("failed to store finalization: %w", err)
 	}
 
+	bl := block.Blacklist()
+	if err := s.db.Put(blacklistKey(bh.Seq), bl.Bytes()); err != nil {
```
Comment (yacovm) [inline simplex/storage.go:181]: I'm OK with this being as it is for now, but we'll eventually have a single metadata object that we'll persist with everything in it.

File: simplex/storage.go [hunk 7]
```diff
@@ -182,7 +195,13 @@ func (s *Storage) Index(ctx context.Context, block simplex.VerifiedBlock, finali
...
 	seqBuff := make([]byte, 8)
 	binary.BigEndian.PutUint64(seqBuff, seq)
-	return seqBuff
+	return append(finalizationPrefix, seqBuff...)
+}
+
+func blacklistKey(seq uint64) []byte {
+	seqBuff := make([]byte, 8)
+	binary.BigEndian.PutUint64(seqBuff, seq)
+	return append(blacklistPrefix, seqBuff...)
```
Comment (yacovm) [inline simplex/storage.go]: same thing. Let's not use this. It's a recipe for improper use. We should allocate an empty slice exactly the size we want and copy into it what we need.

blacklistPrefix is a slice and we might be changing it (in theory) if it has sufficient capacity.
